src/java.base/share/classes/java/lang/template/StringTemplate.java

```diff
@@ -0,0 +1,368 @@
+     */
+    public static int hashCode(StringTemplate stringTemplate) {
+        Objects.requireNonNull(stringTemplate, "stringTemplate should not be null");
+        return Objects.hashCode(stringTemplate.fragments()) ^
+               Objects.hashCode(stringTemplate.values());
+    }
+
+    /**
+     * Tests for equality of two {@link StringTemplate}.
+     *
+     * @param a  first {@link StringTemplate}
+     * @param b  second {@link StringTemplate}
+     *
+     * @return true if the two {@link StringTemplate StringTemplates} are equivalent
+     *
+     * @throws NullPointerException if either @link StringTemplate} is null
+     *
+     * @implSpec Equality is determined by testing equality of the fragments and
+     * the values.
+     */
+    public static boolean equals(Object a, Object b) {
+        Objects.requireNonNull(a, "StringTemplate a should not be null");
+        Objects.requireNonNull(b, "StringTemplate b should not be null");
+        return a instanceof StringTemplate aST && b instanceof StringTemplate bST &&
+               Objects.equals(aST.fragments(), bST.fragments()) &&
+               Objects.equals(aST.values(), bST.values());
+    }
+
+    /**
+     * Returns a StringTemplate composed from a string.
+     *
+     * @param string  single string fragment
+     *
+     * @return StringTemplate composed from string
+     *
+     * @throws NullPointerException if string is null
+     */
+    public static StringTemplate of(String string) {
+        Objects.requireNonNull(string, "string must not be null");
+        return new SimpleStringTemplate(List.of(string), List.of());
+    }
+
+    /**
+     * Returns a StringTemplate composed from fragments and values.
+     *
+     * @implSpec The {@code fragments} list size must be one more that the
+     * {@code values} list size.
+     *
+     * @param fragments list of string fragments
+     * @param values    list of expression values
+     *
+     * @return StringTemplate composed from string
+     *
+     * @throws IllegalArgumentException if fragments list size is not one more
+     *         than values list size
+     * @throws NullPointerException if fragments is null or values is null or if any fragment is null.
+     *
+     * @implNote Contents of both lists are copied to construct immutable lists.
+     */
+    public static StringTemplate of(List<String> fragments, List<Object> values) {
+        Objects.requireNonNull(fragments, "fragments must not be null");
+        Objects.requireNonNull(values, "values must not be null");
+        if (values.size() + 1 != fragments.size()) {
+            throw new IllegalArgumentException(
+                    "fragments list size is not one more than values list size");
+        }
+        for (String fragment : fragments) {
+            Objects.requireNonNull(fragment, "fragments elements must be non-null");
+        }
+        fragments = Collections.unmodifiableList(new ArrayList<>(fragments));
```

Comment:
I think using `List.copyOf()` is more efficient that `Collections.unmodifiableList(new ArrayList<>(...))` because there is no copy if the list is already created with List.of().

Reply:
Changing as recommended